examples/pandas/materialization/my_script.py

```diff
@@ -0,0 +1,49 @@
+# import importlib
+import logging
+import sys
+
+import pandas as pd
+
+from hamilton import driver
+from hamilton.plugins.pandas_extensions import PandasPickleReader, PandasPickleWriter
+
+logging.basicConfig(stream=sys.stdout)
+initial_columns = {  # load from actuals or wherever -- this is our initial data we use as input.
+    # Note: these values don't have to be all series, they could be a scalar.
+    "signups": pd.Series([1, 10, 50, 100, 200, 400]),
+    "spend": pd.Series([10, 10, 20, 40, 40, 50]),
+}
+# we need to tell hamilton where to load function definitions from
+# programmatic code to load modules:
+# module_name = "my_functions"
+# my_functions = importlib.import_module(module_name)
+# or import module(s) directly:
+
+
+import my_functions
+
+dr = driver.Driver(initial_columns, my_functions)  # can pass in multiple modules
+# we need to specify what we want in the final dataframe. These can be string names, or function references.
+output_columns = [
+    "spend",
+    "signups",
+    my_functions.avg_3wk_spend,  # could just pass "avg_3wk_spend" here
+    my_functions.spend_per_signup,  # could just pass "spend_per_signup" here
+    my_functions.spend_zero_mean_unit_variance,  # could just pass "spend_zero_mean_unit_variance" here
+]
+# let's create the dataframe!
+df = dr.execute(output_columns)
+
+print("============ DAG pre-pickle ============")
+print(df.to_string())
+
+# now we pickle the dataframe
+save_path = "df.pkl"
+writer = PandasPickleWriter(path=save_path)
+writer.save_data(df)
+
+reader = PandasPickleReader(filepath_or_buffer=save_path)
+read_df, metadata = reader.load_data(type(df))
+
+print("============ DAG post-pickle ============")
+print(read_df.to_string())
```

Comment:
Almost there. 

What we want to show here instead is invoking execution via the `.materialize` function:
```python
  materializers = [
        # materialize the dataframe to a pickle file
        to.pickle(
            dependencies=output_columns, 
            id="df_to_pickle", 
            path="./data/df.pkl",
            combiner=base.PandasDataFrameResult()  # add from hamilton import base
         ),
    ]
    dr.visualize_materialization(
        *materializers,
        output_file_path="./dag",
        render_kwargs={},
    )
    materialization_results, _ = dr.materialize(
        # materialize model parameters to json
        *materializers,
    )

```
Under the hood, the classes you created will be registered, and the `to.pickle` will magically invoke your code, because it depends on getting a PandasDataframe -- which the `combiner` will produce.


E.g. see
https://github.com/DAGWorks-Inc/hamilton/blob/main/examples/materialization/run.py#L56-L85

Reply:
Hey there, I've done the edits and am getting some errors which I'm struggling to overcome with `visualise_materialization`. 

The error is 
```
(venv) (base) benjaminhack@MacBook-Air hamilton % python examples/pandas/materialization/my_script.py
WARNING:hamilton.telemetry:Note: Hamilton collects completely anonymous data about usage. This will help us improve Hamilton over time. See https://github.com/dagworks-inc/hamilton#usage-analytics--data-privacy for details.
Traceback (most recent call last):
  File "/Users/benjaminhack/Desktop/Ben Hack/PortfolioProjects/hamilton/examples/pandas/materialization/my_script.py", line 51, in <module>
    dr.visualize_materialization(
  File "/Users/benjaminhack/Desktop/Ben Hack/PortfolioProjects/hamilton/hamilton/driver.py", line 1019, in visualize_materialization
    function_graph = materialization.modify_graph(self.graph, materializers)
  File "/Users/benjaminhack/Desktop/Ben Hack/PortfolioProjects/hamilton/hamilton/io/materialization.py", line 211, in modify_graph
    additional_nodes.extend(materializer.resolve(fn_graph))
  File "/Users/benjaminhack/Desktop/Ben Hack/PortfolioProjects/hamilton/hamilton/io/materialization.py", line 109, in resolve
    raise ValueError(
ValueError: Must specify result builder if the materializer has more than one dependency it is materializing. Otherwise we have no way to join them before storage! See materializer df_to_pickle.
```

I've tried adding an adapter to the driver with 

```python
df_builder = base.PandasDataFrameResult()
adapter = base.SimplePythonGraphAdapter(df_builder)

dr = driver.Driver(initial_columns, my_functions, adapter=adapter)
```

as well as adding the output column as an additional var as I think that's whats happening in [https://github.com/DAGWorks-Inc/hamilton/blob/main/examples/materialization/run.py#L56-L85](https://github.com/DAGWorks-Inc/hamilton/blob/main/examples/materialization/run.py#L56-L85). Wish I could be more helpful but am struggling to resolve this error.

Also, I am not being allowed to commit. I need to import the pickling class but then the pre-commit fails because it is not being used. How can I bypass this?

Reply:
Can you push your updated code? I'm not seeing it.

Reply:
You have to pass it as part of the `combiner` argument. Not as an adapter. :)

Reply:
I will try to get some code to show you today/tomorrow. Maybe even directly add it to this PR. We're super close and thanks for all the work. We should be able to finish this, this week!